### ISSUE TITLE
Attempted Edit

### ISSUE THREAD

**Issue by AndreaSchecker** [intersystems-community/vscode-objectscript#781]
Issue Type: <b>Bug</b>

I programmed some server command by my own. That server commands worked fine for a long time.
Suddonly I received 
Studio Action "Attempted Edit" not supported on 127.0.0.1:57772[CTAE01]
by using one of my server commands.
Explanation:
Error running command vscode-objectscript.serverCommands.contextOther: Running the contributed command: 'vscode-objectscript.serverCommands.contextOther' failed.. This is likely caused by the extension that contributes vscode-objectscript.serverCommands.contextOther.

Extension version: 1.1.1
VS Code version: Code 1.61.2 (6cba118ac49a1b88332f312a8f67186f7f3c1643, 2021-10-19T14:57:20.575Z)
OS version: Windows_NT x64 10.0.19042
Restricted Mode: No

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i7-7500U CPU @ 2.70GHz (4 x 2904)|
|GPU Status|2d_canvas: enabled<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>oop_rasterization: enabled<br>opengl: enabled_on<br>rasterization: enabled<br>skia_renderer: enabled_on<br>video_decode: enabled<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|undefined|
|Memory (System)|31.90GB (19.14GB free)|
|Process Argv|. --crash-reporter-id 9812a2f4-d7e0-49de-ac19-9cf9beb0ed10|
|Screen Reader|no|
|VM|0%|
</details><details>
<summary>A/B Experiments</summary>

```
vsliv368:30146709
vsreu685:30147344
python383cf:30185419
vspor879:30202332
vspor708:30202333
vspor363:30204092
vswsl492cf:30256860
pythontb:30283811
pythonptprofiler:30281270
vshan820:30294714
vstes263:30335439
pythondataviewer:30285071
vscod805cf:30301675
pythonvspyt200:30340761
binariesv615:30325510
bridge0708:30335490
dockerwalkthru:30377721
bridge0723:30353136
pythonrunftest32:30373476
pythonf5test824:30373475
javagetstartedt:30391933
pythonvspyt187:30373474
vsqsis200cf:30402084
vsaa593:30376534
vssld246cf:30401891
vscexrecpromptt3:30400063
vscop804:30404766

```

</details>

**Comment by isc-bsaviano** [comment]
@AndreaSchecker How are you executing this server command? Are you right-clicking on an element in the ObjectScript Explorer?

**Comment by AndreaSchecker** [comment]
right clicking on an Element in the ObjectScript Explorer -> Server Command Menue -> Choosing an own command 
(embeded in an Subclass of %Atelier.SourceControl
		} elseif ..actualMenuItem="ITEM10" {
			If Answer=1 {
				do ..log(Msg)
				for x="*.cls","*.int","*.mac","*.inc" {
					Do $system.OBJ.ImportDir(Msg,x,"bck",,1)
				}
				
			}
Output:
Studio Action "Attempted Edit" not supported on 127.0.0.1:57772[CTAE01]
Error running command vscode-objectscript.serverCommands.contextOther: An unknown error occurred. Please consult the log for more details.. This is likely caused by the extension that contributes vscode-objectscript.serverCommands.contextOther.

**Comment by isc-bsaviano** [comment]
@AndreaSchecker I think the error message you're seeing is a little misleading. The issue is that the REST request that's executing your Studio Action is returning an error status. I just opened a PR that will log that error message, which should make diagnosing your issue easier. You could also  use Wireshark or turn on the [HTTP Trace for your server](https://docs.intersystems.com/irislatest/csp/docbook/DocBook.UI.Page.cls?KEY=GCGI_oper_config#GCGI_viewtrace) in the Web Gateway portal to log the REST request and response when you run your Studio Action.

**Comment by AndreaSchecker** [comment]
Sorry, I'm not able to use Wireshark (not allowed by my department) and the link 'HTTP trace for your server' leeds to an IRIS Version. I'm still using Caché for Windows 2018.1.2

**Comment by isc-bsaviano** [comment]
@AndreaSchecker [Here's](https://docs.intersystems.com/ens201815/csp/docbook/DocBook.UI.Page.cls?KEY=GCGI_oper_config#GCGI_viewtrace) the Cache documentation. You can also download and install the beta version of the extension found [here](https://github.com/intersystems-community/vscode-objectscript/releases/download/v1.1.2-beta.8/vscode-objectscript-1.1.2-beta.8.vsix) and try executing your Studio Action again. There should be more information logged in the Output channel than what we saw before.

**Comment by AndreaSchecker** [comment]
Log (Windows)
[2021-12-02 15:24:21.159] [renderer1] [info] Starting extension host with pid 14332.
[2021-12-02 15:24:21.160] [renderer1] [info] [LocalProcessExtensionHost]: IExtensionHostStarter.start() took 110 ms.
[2021-12-02 15:24:25.800] [renderer1] [error] [Extension Host] (node:14332) [DEP0005] DeprecationWarning: Buffer() is deprecated due to security and usability issues. Please use the Buffer.alloc(), Buffer.allocUnsafe(), or Buffer.from() methods instead.
(Use `Code --trace-deprecation ...` to show where the warning was created)

**Comment by AndreaSchecker** [comment]
Log (Settings Sync)
[2021-12-02 15:24:23.556] [settingssync] [info] Using settings sync service https://vscode-sync.trafficmanager.net/
[2021-12-02 15:24:23.556] [settingssync] [info] Auto Sync is disabled.

**Comment by isc-bsaviano** [comment]
@AndreaSchecker Sorry, I meant the `ObjectScript` Output channel. The error should start with "Executing Studio Action ..."

**Comment by AndreaSchecker** [comment]
Executing Studio Action "Attempted Edit" on 127.0.0.1:57772[CTAE01] failed.

**Comment by isc-bsaviano** [comment]
@AndreaSchecker If the Atelier API returned an error message it would have been logged following that line. Would it be possible for you to turn HTTP trace on and log the REST request and response?

**Comment by AndreaSchecker** [comment]
I installed Wireshark and now there are two logfiles. But I have no idea, how I can send them.
I'm trying to turn on HTTP trace, but It's my first time to use CSP (Web Gateway Management) I need a little more time.

**Comment by isc-bsaviano** [comment]
@AndreaSchecker In the wireshark logs, look for a request CTAE01/action/query POST request to the server at 127.0.0.1:57772. The body should contain a JSON object with query `select * from %Atelier_v1_Utils.Extension_UserAction(?, ?, ?, ?)` and "Attempted Edit" in the parameters array. When you click on that request, Wireshark should show you the response.

**Comment by isc-bsaviano** [comment]
You can also save your Wireshark trace as a txt file and add it to your next message and I will look through it for you.

**Comment by AndreaSchecker** [comment]
[WireSharkProtokolle.zip](https://github.com/intersystems-community/vscode-objectscript/files/7662503/WireSharkProtokolle.zip)
Here are two Wireshark-Captures. For I have not used Wireshark before and I don't know, which protokoll ist importent for you,, I zipped the whole Files.

**Comment by isc-bsaviano** [comment]
@AndreaSchecker Thank you for sending me the trace, it's exactly what I needed. I've looked through it and I don't see any errors coming from the server. This means that the failure is somewhere in the extension. Based on the command that you selected, I'd expect that you'd be shown an input box with "Eingabe des kompletten Pfad zum Projektdirectories fÃ¼r CTAE01." as the prompt. Do you get an error before or after this input box is shown?

**Comment by AndreaSchecker** [comment]
I received the error after keying in the access path.

**Comment by isc-bsaviano** [comment]
@AndreaSchecker Ok, so just to to clarify, you were shown the input box, you entered in some data, you submitted it, and then you got the error? If so, I can narrow down where the bug might be

**Comment by AndreaSchecker** [comment]
That's correct.

**Comment by isc-bsaviano** [comment]
@AndreaSchecker Thanks for confirming. After you submit your text, do you see a progress notification titled `Executing AfterUserAction: <your command>`? It may appear very briefly and it will look similar to the one that appears after you selected your command from the menu (before the input box is shown).

**Comment by AndreaSchecker** [comment]
I don't notice such a notification and in output chanel **Executing Studio Action "Attempted Edit" on 127.0.0.1:57772[CTAE01] failed.** appear. 

And I got a new output chanel Log (Extension host):
[2021-12-07 09:46:20.329] [exthost] [error] undefined vscode-objectscript.serverCommands.contextOther {"value":"intersystems-community.vscode-objectscript","_lower":"intersystems-community.vscode-objectscript"}
[2021-12-07 09:46:35.122] [exthost] [error] undefined vscode-objectscript.serverCommands.contextOther {"value":"intersystems-community.vscode-objectscript","_lower":"intersystems-community.vscode-objectscript"}
[2021-12-07 09:46:58.878] [exthost] [error] undefined vscode-objectscript.serverCommands.contextOther {"value":"intersystems-community.vscode-objectscript","_lower":"intersystems-community.vscode-objectscript"}

**Comment by isc-bsaviano** [comment]
@AndreaSchecker It wouldn't appear in the output channel. It would appear in a message box in the bottom right portion of your screen and would disappear once the action was completed on the server. Since an AfterUserAction query didn't appear in your trace, I assume that the failure happened before that point. I will keep looking.

**Comment by isc-bsaviano** [comment]
@AndreaSchecker Do you mind doing another Wireshark trace while you attempt the server action, but this time let it capture packets for a minute or so after you see the error, just to be sure that the AfterUserAction query is captured if it was sent? I think it's very unlikely that the command failed in between you submitting your input and the AfterUserAction query being sent.

**Comment by AndreaSchecker** [comment]
I received a new version of objectscript extension (Ver 1.2.1). Now the command stopped at the checkbox. I can key text in, but after sending return, The box doesn't disappear and the cursor remains in the box. There are no messages down to the right. I clicked outside the input box and the input box vanished.
Therefore I tried another server command menue without a input box. This command was executed, the notification "Executing ..." appeared and then the known error was shown. This actions were recorded in the Wireshark protokoll.
[protokoll20211297.zip](https://github.com/intersystems-community/vscode-objectscript/files/7669300/protokoll20211297.zip)

**Comment by isc-bsaviano** [comment]
@AndreaSchecker The newest version of the extension is actually 1.2.2. I'm not sure why the input box is not accepting your text. Do you see an error appear when you press enter, or does the box just turn red with no messages anywhere? Everything in that trace looks fine. Did you see the "HHLA Hook" console output written to the `ObjectScript` Output channel?

**Comment by AndreaSchecker** [comment]
Now I'm working with the newist version (1.2.2). The problem with the input box didn't occur again.
But I still receive 'Executing Studio Action "Attempted Edit" on 127.0.0.1:57772[TSTCBW] failed.' when I try to perform a server command menue.
Is there something that I can check to solve this issue? I urgendly need these commands.

**Comment by isc-bsaviano** [comment]
@AndreaSchecker Is your command failing in the same place (after you enter your path in the input box)? Do any other server commands or source control hooks work? Has the command or server version changed since the last time you remember it working? Does this server command work in Studio (if you have it available)? Sorry to ask so many questions but I want to rule out a bug in your command.

**Comment by AndreaSchecker** [comment]
I checked it today.
The command is failing after entering my path in the input box. The command is not processed.
Here the command:
**for x="*.cls","*.int","*.mac","*.inc" { Do $system.OBJ.ImportDir(Msg,x,"bck",,1) }**
There are other commands without a checkbox, for exp.
**do ..exportFile(InternalName, fileName)**
This command is proceessed before the failure is displayed.

**Comment by isc-bsaviano** [comment]
@AndreaSchecker Thanks for the additional information. I think your issue needs support from the WRC. I suggest that you contact them to open a new case and refer them to this issue for background information.

**Comment by AndreaSchecker** [comment]
I opened wrc 955035

**Comment by GabrielGonzalezRivero** [comment]
I have been working with Andrea in WRC 955035. I wrote a very simple source control hook that reproduces the problem. All it does is store the menu name into a global:
Class Test.SourceControl Extends %Atelier.SourceControl
{

XData Menu
{
<MenuBase>
<Menu Name="MENU" Type="1">
<MenuItem Name="RunTest" Enabled="1"/>
</Menu>
</MenuBase>
}

Method UserAction(Type As %Integer, Name As %String, InternalName As %String, SelectedText As %String, ByRef Action As %String, ByRef Target As %String, ByRef Msg As %String, ByRef Reload As %Boolean) As %Status
{
s ^Gabe = Name
q $$$OK
}

}


When you right click any routine in the 'Objectscript' view, and select the RunTest menu item I created, the source control code runs completion without any problem, and you can even check the ^Gabe global and see the name of the menuitem was stored, but then you get the following error in VSCode:
*** Atelier Source Control Hook *** OnMenuItem MENU,RunTest EnsJob.MAC  1 RunTest
Executing Studio Action "Attempted Edit" on 127.0.0.1:52773[USER] failed.

This seems to only happen in the Objectscript view. If you add a server namespace to the workspace directly,  and do the same steps from the Explorer view, it runs without error.
I have everything running on latest:
Internal Language Server 1.2.6
Intersystems Objectscript 1.2.2
Intersystems Objectscript Extension Pack 1.0.3
Intersystems Server Manager 2.0.6
IRIS for Windows (x86-64) 2021.1 (Build 215U) Wed Jun 9 2021 09:39:22 EDT

Furthermore, it is unclear what this "Attempted Edit" studio action is. I searched the string "Attempted Edit" in all of the inherited classes, and none of them had any hits. To me this looks like the VSCode side must be trying to run somethingcalled "Attempted Edit" after the source control hook code is done executing the server command.

**Comment by isc-bsaviano** [comment]
@GabrielGonzalezRivero Thanks for looking into this. I downloaded your test class and made it my active source control class. I was able to reproduce the behavior you described. It turns out that a simple typo was causing Studio Actions that required no user input to be treated as `Attempted Edit` Studio Actions. I opened a PR that addresses this issue. You and @AndreaSchecker can test the fix by downloading and installing the dev vsix found [here](https://github.com/intersystems-community/vscode-objectscript/suites/4694540603/artifacts/128209288).

**Comment by AndreaSchecker** [comment]
@isc-bsaviano I checked your fix and it worked fine. Then I received a vscode update to Version 
_Version: 1.63.2 (system setup)
Commit: 899d46d82c4c95423fb7e10e68eba52050e30ba3
Date: 2021-12-15T09:40:02.816Z
Electron: 13.5.2
Chromium: 91.0.4472.164
Node.js: 14.16.0
V8: 9.1.269.39-electron.0
OS: Windows_NT x64 10.0.19042_
Now I receive when I trie to use Server Command Menue
<img width="271" alt="image" src="https://user-images.githubusercontent.com/94963324/146757878-adae14ee-4624-4197-9f57-11a57a4b754a.png">
The Server Command Menue does not open.
In your vsix File there is no vscode-objectscript.serverCommands.contextOther
The aktual objectscript version V1.2.2 works usually

**Comment by isc-bsaviano** [comment]
@AndreaSchecker I installed VS Code 1.63.2 and extension version 1.2.3-dev.810 and the fix works using @GabrielGonzalezRivero's test source control class. Can you work with him to test the fix using your source control class?

**Comment by AndreaSchecker** [comment]
@isc-bsaviano  Would you be so kind to check whether your hotfix works under VSC Version 1.6.3?
Under Version 1.6.2 it works fine, but then I got an upgrade and then I even can't open Server Control Menue.
I need the fix vera urgently

**Comment by isc-bsaviano** [comment]
@AndreaSchecker I'm not sure what version of VS Code you're talking about. I have the latest version (1.63.2) and the vsix with the fix works on that version.. There is no version 1.63.3. Are you using VS Code Insiders? That would be version 1.64.0-insiders

**Comment by AndreaSchecker** [comment]
@isc-bsaviano : You are right I use Version 1.6.2. But this Version reacts like I mentioned for Version 1.63.3. When I use your hotfix there are no Server Command menue any more. I have no own commands to choose.

**Comment by isc-bsaviano** [comment]
@AndreaSchecker I'm using VS Code version 1.63.2 and vscode-objectscript version 1.2.3-dev.810. I can successfully run the test action provided by Gabriel's test class and can run the source control actions provided by ISC's source control class. I think you should keep working with Gabriel because I can't reproduce this new issue.

**Comment by GabrielGonzalezRivero** [comment]
@isc-bsaviano I opened my VSCode running objectscript 1.2.2 and made sure I could still reproduce the problem. I then upgraded the objectscript extension to the 1.2.3-dev.810 version you provided and now when I right click a class in the objectscript view and select 'Server command Menu', and it no longer prompts me for the server command I want to run. I simply get this error:
Error running command vscode-objectscript.serverCommands.contextOther: command 'vscode-objectscript.serverCommands.contextOther' not found. This is likely caused by the extension that contributes vscode-objectscript.serverCommands.contextOther.

I tried completely uninstalling the objectscript extension (and the language server since it depends on it), and reinstalling it using the file you provided, and this does not change the behavior. If I revert back to objectscript 1.2.2, I get the original behavior again.

**Comment by GabrielGonzalezRivero** [comment]
I worked with @isc-bsaviano , and we discovered why its not prompting for the server command. The issue is that the patch is not actually running, so the contextOther command doesn't exist. Upon testing we found that the problem wasn't limited to just this patch, but it seems there is trouble loading several of the beta patches as well. 

Upon doing some debugging we found that the extension was failing to activate because of this error:
CRITI Extension 'intersystems-community.vscode-objectscript CANNOT USE these API proposals 'fileSearchProvider, textSearchProvider'. You MUST start in extension development mode or use the --enable-proposed-api command line flag

This error happens when something is  tryin to use the proposed-api code, but the flag for enabling the proposed-api code is not set. Brett was able to track down what is trying to use this code. He will check in a change to not require this flag to be set (dev814), and once this is merged, he will update the dev810 patch.

In the mean time, the work around this problem, we can simply enable the proposed-api flag. To do this:
1) Open VSCode and press ctl+shift+P to open the Cmmand Palette
2) choose 'Preferences: Configure Runtime Arguments'
3) In the argv.json file that opens, add this line:
"enable-proposed-api": ["intersystems-community.vscode-objectscript"],

4) exit VSCode and relaunch it. Once you go to the objectscript explorer view, (and enter your username/password if prompted), you should see the following message in the objectscript output window:
intersystems-community.vscode-objectscript version 1.2.3-dev.810 activating with proposed APIs available.

The dev810 patch should now work, and therefore the source control server commands as well.

**Comment by AndreaSchecker** [comment]
I enabled the proposed-api flag and now everthing works fine.
Thanks to Brett and Gabriel.
